### ISSUE TITLE
v3.7 - booth CMake and configure for MinGW fails

### ISSUE THREAD

**Issue by mmidor** [cpputest/cpputest#628]
Error when compiling /Platforms/Gcc/UtestPlatform.cpp:
functions PlatformSpecificForkImplementation() and PlatformSpecificWaitPidImplementation() are incorrectly covered by #ifdef __MINGW32__.

Tested with v3.6 - configuration works, because /Platforms/Gcc/UtestPlatform.cpp looks different.
My package comes from http://cpputest.github.io/, cpputest-3.7.zip

**Comment by arstrube** [comment]
You are right -- they need to be in the #ifndef part. I'll fix it, of course. But it is too bad this was only noticed after the release, so now it is wrong in 3.7......

@Andne would it be possible to test a MinGW build on AppVeyor?

**Comment by Andne** [comment]
In theory, yes, AppVeyor does have MinGW listed in their supported software list.  In practice, I'm going to have to try things a bit to see how it interacts with the current build matrix.

**Comment by arstrube** [comment]
Even a simple "quick" build would be great, because it seems none of our 'regulars' use MinGW since I switched to Cygwin, so problems like this one tend to go undetected.

**Comment by Andne** [comment]
Will using CMake 2.8 to run the build work?

**Comment by arstrube** [comment]
I guess it should.

**Comment by Andne** [comment]
The good news is I can invoke both CMake and MinGW on appveyor, the bad news is that the copy of sh.exe that git uses is causing CMake to be unhappy.

**Comment by arstrube** [comment]
With MinGW, the 'rest' of Linux commands -- including make -- live in msys. Git, on the other hand, seems to use a stripped-down version of msys, which does not include make. At least on VM. Exactly what causes CMake to be unhappy?

**Comment by arstrube** [comment]
@Andne - I finally managed a MinGW build, like this:
1. Download MinGW make and copy to Git/bin
2. Git Bash here (in CppUTest root)
3. Set PATH appropriately
4. make -f Makefile_using_MakefileWorker
I couldn't make cmake happy whatever I did (it seems to expect stuff in strange places if you specify "MinGW Makefiles".
This just builds and tests CppUTest, not the extensiosns. I'm sure that can be fixed giving it the right target(s).
So, in short, building using the old makefile rather than cmake seems to work. Maybe that would wokr on AppVeyor, too?

**Comment by arstrube** [comment]
```bash
$ make -f Makefile_using_MakefileWorker all extensions
```
should do the trick

**Comment by basvodde** [comment]
What about configure. That should work?

**Comment by arstrube** [comment]
Would depend on Appveyor having automake. I wouldn't bank on that, even if it does have MinGW.

**Comment by arstrube** [comment]
On my machine, automake is part of standard msys, but the whole thing is jinxed somehow. (all msys commands just do nothing).

**Comment by arstrube** [comment]
MinGW separates the compiler stuff from the general Linux stuff (which is msys). So having the compiler doesn't per se mean having msys. Git, on the other hand, has the essentials of msys, but not make and certainly not automake.

**Comment by basvodde** [comment]
Yeah, I forgot that we don't put the configure file in github anymore :)

**Comment by arstrube** [comment]
I thought I posted two links to Appveyor here that give some clues to how it can be done. But they appeat to be gone ?

**Comment by Andne** [comment]
You posted them in the .gitattribute issue.

I managed to add logic to strip the git folders from the path, so now CMake is happy with that part of things.  I'm still working on getting it to actually run - complains that the compiler isn't a working compiler at the moment.

**Comment by arstrube** [comment]
Oops. Stupid me. Got to the right person anyway ;-). Could you give the old makefile a try? I think Bas is right in that autotools should work, too. But that would probably be more work.

**Comment by Andne** [comment]
Well, I got cmake to run correctly on appveyor, granted, the build fails because of the bug that started this.
https://ci.appveyor.com/project/Andne/cpputest/build/job/0x4eig5voq3lpt8f

Do we want to add an autotools and/or old makefile build as well?

**Comment by arstrube** [comment]
That's great. I don't really care which version of make, I think one is enough. Did you run your test off the latest master? It shouldn't have failed anymore

**Comment by Andne** [comment]
No, the branch was made yesterday when I first started working on it, so it didn't have the other changes in it.  I've submitted pull request #631 that should use the latest sources for it's build.

**Comment by arstrube** [comment]
Oh wait. I'ts pthread. autotools would probably catch this and set a flag.

It isn't that MinGW doesn't have pthread support per se. But the installation on Appveyor doesnt seem to have it

**Comment by arstrube** [comment]
That wasn't very clear. Whereas MinGW never has fork(), many distributions do have posix thread support.

**Comment by arstrube** [comment]
Since GCC 4.8.1 there are several flavors of MinGW, that support either Posix threads or Windows threads. This is probably the "wrong" one. Rather than change CppUTest from what it has always been, can we perhaps install the correct version of MinGW onto the build server (this is what others seem to be doing, although a bit of a bother. And as an aside, the reason why I personally switched from MinGW to Cygwin ;-).

**Comment by arstrube** [comment]
For example, this one:

http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win32/Personal%20Builds/mingw-builds/4.8.3/threads-posix/dwarf/i686-4.8.3-release-posix-dwarf-rt_v3-rev2.7z/download

And yes, one has to choose between dwarf and sjlj as well... :-(

**Comment by arstrube** [comment]
Chocolatey might do the trick -- see here:

https://github.com/ferventcoder/chocolatey-packages/blob/master/manual/mingw/mingw.nuspec

**Comment by Andne** [comment]
Ok, so the build ran, but failed due to missing pthread.h.  I think that shows us that the version of MinGW that appveyor has by default does not support pthreads, and so can't use fork and such. Is that what is desired?

**Comment by arstrube** [comment]
Err, no fork() is a given. That's okay. No pthread needs to be fixed on Appveyor somehow :-)

**Comment by arstrube** [comment]
Essentially we need to :
1. Get a version of MinGW that has Posix threads (via Chocolatey?)
2. Install that into, say C:\MinGW493_32.
3. Take out C:\MinGW from path and put in C:\MinGW493_32
4. Run the build normally
I'd recommend gcc 4.9.3 since I believe that's what's used on TravisCI
@bortsov - can you help out with 1) maybe?
@Andne - aside from the pthread issue the pull looks good, many thanks!
